File: backend/services/exercise_tracking_metric.py
"""
Exercise tracking-metric derivation (the single source of truth).

Cardio / functional / timed exercises (SkiErg, Sled Push/Pull, Farmers Carry,
Rowing/RowErg, runs, Plank Hold, Wall Balls, Burpees…) must NOT render in the
active-workout UI as "weight × reps". The workout payload historically carried
no structured tracking metadata, so the Flutter client fell back to a phantom
10 kg × reps card and ``num.tryParse("1000 m") -> null`` dropped the unit.

This module derives, per exercise, a ``tracking_type`` enum + the structured
target (distance / duration) the client needs, plus a raw unit-bearing
``reps_spec`` string as a last-resort display/parse fallback.

Pure + deterministic + dependency-free (no DB, no network) so it can be:
  • applied at every serve-time chokepoint (``/today``, workout-detail,
    program expansion) with zero added latency, and
  • unit-tested in isolation.

The decision is sourced FIRST from canonical ``exercise_library`` metadata
(``equipment``, ``is_timed``, ``movement_pattern`` — when present on the
exercise dict or passed as ``library_meta``), then falls back to parsing the
program's reps / reps_spec string + the exercise name. This mirrors the
fail-open, name-classifier philosophy of ``_attach_movement_meta`` in
``api/v1/workouts/today.py``: an unknown exercise yields ``tracking_type=None``
(the frontend has its own classifier), never a wrong default.
"""
from __future__ import annotations

import re
from typing import Any, Dict, List, Optional


# ---------------------------------------------------------------------------
# Tracking-type enum (string literals shared with the Flutter contract)
# ---------------------------------------------------------------------------
TRACK_WEIGHT = "weight"
TRACK_BODYWEIGHT = "bodyweight"
TRACK_TIME = "time"
TRACK_DISTANCE = "distance"


# ---------------------------------------------------------------------------
# Name lexicons (lowercase substring match). Kept deliberately specific — a
# false positive here mislabels a real loaded lift, so we only list movements
# whose PRIMARY metric is unambiguous.
# ---------------------------------------------------------------------------

# Distance-primary: cardio machines, sleds, loaded carries, measured runs,
# horizontal jumps. These may ALSO be timed; distance wins for the primary
# metric but we still surface duration when present.
_DISTANCE_NAME_HINTS = (
    "skierg", "ski erg", "ski-erg",
    "rowerg", "row erg", "row-erg", "rowing machine", "erg row",
    "rowing", "row machine",
    "sled push", "sled pull", "sled drag", "sled sprint", "prowler",
    "farmer", "farmers carry", "farmer's carry", "farmers walk",
    "suitcase carry", "suitcase walk", "loaded carry", "yoke carry",
    "yoke walk", "sandbag carry", "overhead carry", "waiter walk",
    "broad jump", "standing long jump",
    "shuttle run", "shuttle sprint", "beep test",
    "run", "jog", "sprint", "walk lunge distance",
    "bike erg", "bikeerg", "assault bike distance",
)

# Names that look like a carry/run but are NOT distance-tracked, so the broad
# hints above never mislabel them. (e.g. "carry" inside "farmers carry hold").
_DISTANCE_NAME_EXCLUDE = (
    "calf raise", "farmer hold",  # static holds, not measured by distance
)

# Bodyweight rep movements (equipment-free, rep-counted). Wall Ball is included
# per the product contract even though it uses a med ball — its load is fixed
# and the user counts reps, not weight.
_BODYWEIGHT_NAME_HINTS = (
    "burpee", "air squat", "bodyweight squat", "push-up", "push up", "pushup",
    "pull-up", "pull up", "pullup", "chin-up", "chin up", "chinup",
    "mountain climber", "jumping jack", "box jump", "tuck jump", "jump squat",
    "squat jump", "wall ball", "wall-ball", "sit-up", "sit up", "situp",
    "v-up", "v up", "toes to bar", "toes-to-bar", "knees to elbow",
    "inverted row", "muscle-up", "muscle up", "high knees", "butt kick",
    "bear crawl", "crab walk", "skater", "star jump",
)

# Timed holds (isometric / duration-based). The user tracks seconds, not reps.
_TIME_NAME_HINTS = (
    "plank", "wall sit", "wall-sit", "dead hang", "dead-hang", "bar hang",
    "hollow hold", "hollow body hold", "l-sit", "l sit", "side plank",
    "superman hold", "glute bridge hold", "hold", "isometric",
    "static hold", "wall squat hold", "boat pose", "chair pose",
)

# Bodyweight equipment tokens — equipment values that mean "no external load".
_BODYWEIGHT_EQUIPMENT = {"bodyweight", "body weight", "none", "no equipment", ""}

# Movement patterns (from services.exercise_rag.sane_ranges) that are
# rep-based bodyweight movements.
_BODYWEIGHT_PATTERNS = {
    "bodyweight_push", "bodyweight_pull", "bodyweight_squat",
}
_CARRY_PATTERNS = {"carry"}

# Loaded sleds / carries: a distance/time station that ALSO takes a load, so its
# metric set includes `weight`. Mirrored in the Flutter classifier
# (ExerciseTrackingMetric.isLoadedCarry).
_LOADED_CARRY_HINTS = (
    "sled", "prowler", "yoke", "farmer", "suitcase carry", "loaded carry",
    "overhead carry", "waiter walk", "sandbag carry", "sandbag lunge",
)
_LOADED_EQUIPMENT = {
    "sled", "sandbag", "dumbbell", "dumbbells", "kettlebell", "kettlebells",
    "barbell", "weight plate", "trap bar", "yoke",
}


def _is_loaded_carry(name: str, equipment_tokens: List[str]) -> bool:
    if _name_matches(name, _LOADED_CARRY_HINTS):
        return True
    return any(t in _LOADED_EQUIPMENT for t in equipment_tokens)


def _metric_keys_for(result: Dict[str, Any], exercise: Dict[str, Any],
                     library_meta: Optional[Dict[str, Any]]) -> Optional[List[str]]:
    """Ordered metric columns for an exercise, from its primary tracking_type +
    the loaded-carry rule. weight lift -> [weight,reps]; bodyweight -> [reps];
    time -> [time]; distance -> [distance]; loaded carry -> weight prepended."""
    tt = result.get("tracking_type")
    if tt == TRACK_WEIGHT:
        keys = ["weight", "reps"]
    elif tt == TRACK_BODYWEIGHT:
        keys = ["reps"]
    elif tt == TRACK_TIME:
        keys = ["time"]
    elif tt == TRACK_DISTANCE:
        keys = ["distance"]
    else:
        return None
    if tt in (TRACK_DISTANCE, TRACK_TIME) and _is_loaded_carry(
        _name_of(exercise), _equipment_tokens(exercise, library_meta)
    ):
        keys = ["weight"] + keys
    return keys


# ---------------------------------------------------------------------------
# String parsers (robust, unit-bearing).
# ---------------------------------------------------------------------------

# Distance: "1 km", "1km", "1.5 km", "1000 m", "1000m", "200 meters",
# "400 yards", "400 yd", "1 mile", "1.2 miles", "880 yards".
_DISTANCE_RE = re.compile(
    r"(\d+(?:\.\d+)?)\s*"
    r"(km|kilometers?|kilometres?|m(?:eters?|etres?)?|mi(?:les?)?|yd|yards?)\b",
    re.IGNORECASE,
)

# Duration: "8 minutes", "8 min", "2 min", "30s", "45s hold", "30 sec",
# "90 seconds", "1 hour". NOTE: a bare "m"/"meters" is distance, never minutes,
# so the minute token requires "min". Hour supported for completeness.
_DURATION_RE = re.compile(
    r"(\d+(?:\.\d+)?)\s*"
    r"(h(?:ours?|rs?)?|min(?:utes?|s)?|s(?:ec(?:onds?|s)?)?)\b",
    re.IGNORECASE,
)


def parse_distance_meters(text: Any) -> Optional[float]:
    """Parse a free-text distance target into meters, or None.

    "1 km"->1000, "1000 m"->1000, "200m"->200, "400 yards"->365.76,
    "1 mile"->1609.34. Returns None when no distance unit is present.
    """
    if text is None:
        return None
    if isinstance(text, (int, float)):
        # Bare number with no unit is ambiguous (reps vs meters) — refuse.
        return None
    m = _DISTANCE_RE.search(str(text))
    if not m:
        return None
    value = float(m.group(1))
    unit = m.group(2).lower()
    if unit.startswith("km") or unit.startswith("kilom"):
        return round(value * 1000.0, 2)
    if unit.startswith("mi"):
        return round(value * 1609.34, 2)
    if unit.startswith("yd") or unit.startswith("yard"):
        return round(value * 0.9144, 2)
    # meters family ("m", "meter", "metre", "meters")
    return round(value, 2)


def parse_duration_seconds(text: Any) -> Optional[int]:
    """Parse a free-text duration target into whole seconds, or None.

    "8 minutes"->480, "30s"->30, "45s hold"->45, "2 min"->120, "1 hour"->3600.
    A bare number or a distance unit returns None.
    """
    if text is None:
        return None
    if isinstance(text, (int, float)):
        return None
    m = _DURATION_RE.search(str(text))
    if not m:
        return None
    value = float(m.group(1))
    unit = m.group(2).lower()
    if unit.startswith("h"):
        return int(round(value * 3600))
    if unit.startswith("min"):
        return int(round(value * 60))
    # seconds family
    return int(round(value))


# ---------------------------------------------------------------------------
# Field readers (tolerant of the many shapes an exercise dict takes across the
# AI / curated / custom generation paths).
# ---------------------------------------------------------------------------
def _name_of(ex: Dict[str, Any]) -> str:
    return str(ex.get("name") or ex.get("exercise_name") or "").strip().lower()


def _reps_target_string(ex: Dict[str, Any]) -> Optional[str]:
    """Best raw, unit-bearing target string for display/parse fallback.

    Prefers an explicit reps string; renders a structured ``reps_spec`` dict
    to its human form; tolerates a bare numeric reps value.
    """
    spec = ex.get("reps_spec")
    if isinstance(spec, dict):
        try:
            from services.program_library_importer import reps_spec_display
            rendered = reps_spec_display(spec)
            if rendered:
                return rendered
        except Exception:
            # Fall through to the raw reps below — never raise from a serializer.
            pass
    if isinstance(spec, str) and spec.strip():
        return spec.strip()
    reps = ex.get("reps")
    if reps is None:
        # Time-based holds sometimes carry duration only.
        for k in ("duration", "duration_text", "target"):
            v = ex.get(k)
            if v:
                return str(v)
        return None
    return str(reps)


def _equipment_tokens(ex: Dict[str, Any], library_meta: Optional[Dict[str, Any]]) -> List[str]:
    raw = None
    if library_meta and library_meta.get("equipment") is not None:
        raw = library_meta.get("equipment")
    if raw is None:
        raw = ex.get("equipment")
    tokens: List[str] = []
    if isinstance(raw, str):
        tokens = [raw.strip().lower()]
    elif isinstance(raw, list):
        tokens = [str(t).strip().lower() for t in raw if t is not None]
    return tokens


def _is_bodyweight_equipment(tokens: List[str]) -> bool:
    if not tokens:
        return False
    return all(t in _BODYWEIGHT_EQUIPMENT for t in tokens)


def _name_matches(name: str, hints: tuple) -> bool:
    return any(h in name for h in hints)


# ---------------------------------------------------------------------------
# The single derivation entry point.
# ---------------------------------------------------------------------------
def _derive_core(
    exercise: Dict[str, Any],
    library_meta: Optional[Dict[str, Any]] = None,
) -> Dict[str, Any]:
    """Derive structured tracking metadata for one exercise dict.

    Returns a dict with keys:
      ``tracking_type``   : "weight"|"bodyweight"|"time"|"distance"|None
      ``distance_meters`` : float target meters (distance only) | None
      ``duration_seconds``: int target seconds (when known) | None
      ``hold_seconds``    : int hold seconds (timed holds) | None
      ``reps_spec``       : raw unit-bearing target string | None

    ``library_meta`` (optional) is the canonical ``exercise_library`` row for
    this exercise (``equipment``, ``is_timed``, ``movement_pattern``,
    ``default_hold_seconds``); when supplied it is consulted FIRST.

    Pure + fail-open: any ambiguity that can't be resolved yields
    ``tracking_type=None`` so the client's own classifier decides.
    """
    lm = library_meta or {}
    name = _name_of(exercise)
    reps_spec_str = _reps_target_string(exercise)

    # Corroborating signals, library-first.
    pattern = lm.get("movement_pattern") or exercise.get("movement_pattern")
    is_timed = lm.get("is_timed")
    if is_timed is None:
        is_timed = exercise.get("is_timed")
    hold_seconds = (
        lm.get("default_hold_seconds")
        or exercise.get("hold_seconds")
        or exercise.get("default_hold_seconds")
    )
    duration_seconds = exercise.get("duration_seconds") or lm.get("duration_seconds")
    equipment_tokens = _equipment_tokens(exercise, library_meta)

    # Parse the unit-bearing target string once.
    parsed_distance = parse_distance_meters(reps_spec_str)
    parsed_duration = parse_duration_seconds(reps_spec_str)

    result: Dict[str, Any] = {
        "tracking_type": None,
        "distance_meters": None,
        "duration_seconds": None,
        "hold_seconds": None,
        "reps_spec": reps_spec_str,
    }

    # --- 1) DISTANCE (highest precedence for the primary metric) -------------
    name_is_distance = (
        _name_matches(name, _DISTANCE_NAME_HINTS)
        and not _name_matches(name, _DISTANCE_NAME_EXCLUDE)
    )
    pattern_is_carry = pattern in _CARRY_PATTERNS
    if parsed_distance is not None or name_is_distance or pattern_is_carry:
        result["tracking_type"] = TRACK_DISTANCE
        result["distance_meters"] = parsed_distance
        # Distance stations can also be timed (e.g. "8 minutes" AMRAP row).
        if parsed_duration is not None:
            result["duration_seconds"] = parsed_duration
        elif isinstance(duration_seconds, (int, float)) and duration_seconds:
            result["duration_seconds"] = int(duration_seconds)
        return result

    # --- 2) TIME (isometric holds / duration-based) --------------------------
    name_is_time = _name_matches(name, _TIME_NAME_HINTS)
    has_hold = bool(hold_seconds) or bool(is_timed)
    if parsed_duration is not None or name_is_time or has_hold:
        result["tracking_type"] = TRACK_TIME
        secs = (
            parsed_duration
            if parsed_duration is not None
            else (int(hold_seconds) if hold_seconds else None)
        )
        if secs is None and isinstance(duration_seconds, (int, float)) and duration_seconds:
            secs = int(duration_seconds)
        result["duration_seconds"] = secs
        if hold_seconds:
            result["hold_seconds"] = int(hold_seconds)
        elif secs is not None and (name_is_time or is_timed):
            result["hold_seconds"] = secs
        return result

    # --- 3) BODYWEIGHT (rep-based, no external load) -------------------------
    name_is_bodyweight = _name_matches(name, _BODYWEIGHT_NAME_HINTS)
    pattern_is_bodyweight = pattern in _BODYWEIGHT_PATTERNS
    equip_is_bodyweight = _is_bodyweight_equipment(equipment_tokens)
    if name_is_bodyweight or pattern_is_bodyweight or (
        equip_is_bodyweight and result["reps_spec"] is not None
    ):
        result["tracking_type"] = TRACK_BODYWEIGHT
        return result

    # --- 4) WEIGHT (default: everything else is a loaded lift) ---------------
    result["tracking_type"] = TRACK_WEIGHT
    return result


def derive_tracking_metadata(
    exercise: Dict[str, Any],
    library_meta: Optional[Dict[str, Any]] = None,
) -> Dict[str, Any]:
    """Derive tracking metadata for one exercise: the primary ``tracking_type``,
    structured targets (distance/duration/hold), AND the ordered ``metric_keys``
    column list (the generic capability set — e.g. a loaded sled push yields
    ``["weight","distance"]``). Pure + fail-open. ``library_meta`` is consulted
    first when supplied. See module docstring.
    """
    result = _derive_core(exercise, library_meta)
    keys = _metric_keys_for(result, exercise, library_meta)
    if keys:
        result["metric_keys"] = keys
    return result


# ---------------------------------------------------------------------------
# In-place applier used at the serve-time chokepoints.
# ---------------------------------------------------------------------------
def attach_tracking_metadata(exercises: Any) -> None:
    """In-place: tag each exercise dict in ``exercises`` with tracking metadata.

    Sets ``tracking_type`` + ``distance_meters`` always; backfills
    ``duration_seconds`` / ``hold_seconds`` / ``reps_spec`` only when derived
    and not already present (never clobbers a real generation-time value).

    Cheap + deterministic — safe to call on every workout serialize. No-op for
    non-list / non-dict entries.
    """
    if not isinstance(exercises, list):
        return
    for ex in exercises:
        if not isinstance(ex, dict):
            continue
        try:
            meta = derive_tracking_metadata(ex)
        except Exception:
            # A serializer must never raise — skip a single bad entry.
            continue
        if meta.get("tracking_type"):
            ex["tracking_type"] = meta["tracking_type"]
        if meta.get("metric_keys"):
            ex["metric_keys"] = meta["metric_keys"]
        if meta.get("distance_meters") is not None:
            ex["distance_meters"] = meta["distance_meters"]
        # Backfill duration/hold/reps_spec without overwriting real values.
        if meta.get("duration_seconds") is not None and not ex.get("duration_seconds"):
            ex["duration_seconds"] = meta["duration_seconds"]
        if meta.get("hold_seconds") is not None and not ex.get("hold_seconds"):
            ex["hold_seconds"] = meta["hold_seconds"]
        # reps_spec is the raw unit-bearing string the frontend parses as a last
        # resort. Only set it as a string when it isn't already a non-empty
        # string (preserve any structured value the program path persisted).
        if meta.get("reps_spec") and not (
            isinstance(ex.get("reps_spec"), str) and ex.get("reps_spec").strip()
        ):
            ex["reps_spec"] = meta["reps_spec"]

File: mobile/flutter/lib/core/utils/exercise_tracking_metric.dart
// Exercise tracking-metric classifier.
//
// Decides HOW an exercise is logged — by load, bodyweight reps, time, or
// distance — for ANY exercise: curated-program, AI-generated, or custom. The
// active-workout screens use this to pick the right input UI and to suppress
// the phantom "10 kg × 1" default that cardio / functional / timed / bodyweight
// moves used to show.
//
// Precedence (most-trusted signal first):
//   1. Backend hint  — `tracking_type` emitted by the workout serializer.
//   2. Distance      — a distance machine / carry / sled / run, an explicit
//                      `distance_meters`, or a distance-unit target string.
//   3. Time          — `is_timed` / hold / duration, or a time-unit target.
//   4. Bodyweight    — bodyweight/none equipment, or a known bodyweight rep move.
//   5. Weight        — everything else (loaded lifts).
//
// The frontend lists are a backstop: once the backend emits `tracking_type` /
// `distance_meters` (and the exercise_library carries correct equipment /
// is_timed), (1) short-circuits and this heuristic rarely runs — but it keeps
// the app correct offline and for un-tagged custom moves.

/// The metric an exercise is logged by.
enum TrackingMetric {
  /// Load × reps (barbell, dumbbell, machine, cable).
  weight,

  /// Reps only — no weight column (burpees, push-ups, air squats).
  bodyweight,

  /// A hold / duration timer (plank, wall sit, dead hang, cardio-by-time).
  time,

  /// Distance in meters, optionally with a timer (SkiErg, row erg, sled,
  /// loaded carries, runs).
  distance,
}

extension TrackingMetricX on TrackingMetric {
  bool get isWeight => this == TrackingMetric.weight;
  bool get isBodyweight => this == TrackingMetric.bodyweight;
  bool get isTime => this == TrackingMetric.time;
  bool get isDistance => this == TrackingMetric.distance;

  /// True when the move has NO external load to log (everything except weight).
  bool get hasNoWeight => this != TrackingMetric.weight;
}

/// A parsed target string, e.g. "1000 m" → (distance, 1000, "m").
class TargetSpec {
  final TrackingMetric metric;

  /// Numeric magnitude in the metric's canonical unit:
  ///   distance → meters · time → seconds · bodyweight → reps · weight → reps.
  final num? value;
  const TargetSpec(this.metric, this.value);
}

class ExerciseTrackingMetric {
  ExerciseTrackingMetric._();

  // ── Name fragments (lowercased, substring match) ────────────────────────

  /// Cardio machines + sled + loaded carries + distance running → distance.
  static const List<String> _distanceFragments = [
    'skierg', 'ski erg', 'ski-erg',
    'rowerg', 'row erg', 'rowing machine', 'rowing', 'erg row',
    'sled push', 'sled pull', 'sled drag', 'sled sprint', 'prowler', 'yoke',
    "farmer's carry", 'farmers carry', 'farmer carry', 'farmers walk',
    "farmer's walk", 'loaded carry', 'suitcase carry', 'overhead carry',
    'sandbag carry', 'sandbag lunge', // HYROX sandbag lunges are distance (m)
    'broad jump', 'bear crawl', 'sprint', 'shuttle run', 'beep test',
  ];

  /// Distance-running names that are distance ONLY when measured by distance
  /// (we still let an explicit duration override to time upstream).
  static const List<String> _runFragments = [
    'run', 'jog', 'walk', // qualified below by a distance signal
  ];

  /// Isometric holds / time-based moves.
  static const List<String> _timeFragments = [
    'plank', 'wall sit', 'wall-sit', 'dead hang', 'hollow hold', 'hollow body',
    'l-sit', 'l sit', 'hold', 'isometric', 'static',
    'flutter kick', 'superman hold', 'bridge hold', 'side plank',
  ];

  /// Classic bodyweight rep moves (no external load).
  static const List<String> _bodyweightFragments = [
    'burpee', 'air squat', 'jump squat', 'jumping jack', 'mountain climber',
    'push-up', 'push up', 'pushup', 'pull-up', 'pull up', 'pullup', 'chin-up',
    'chin up', 'sit-up', 'situp', 'sit up', 'crunch', 'box jump', 'tuck jump',
    'high knee', 'butt kick', 'bodyweight squat', 'bodyweight lunge',
    'jumping lunge', 'star jump', 'inchworm', 'bear hold',
    // Med-ball / bar-gymnastics moves logged by reps, not a per-set load
    // (wall balls use a standard fixed med-ball; toes-to-bar is pure
    // bodyweight). Kept narrow on purpose — ambiguous moves that have common
    // weighted variants (dips, chin-ups, thrusters, KB swings) are NOT here so
    // their load is never silently suppressed.
    'wall ball', 'wall-ball', 'toes to bar', 'toes-to-bar', 'knees to elbow',
  ];

  static const List<String> _bodyweightEquipment = [
    'bodyweight', 'body weight', 'none', 'no equipment', 'calisthenics',
  ];

  /// Resolve the tracking metric from the available signals.
  ///
  /// [trackingTypeHint] is the backend `tracking_type` string (preferred).
  /// [distanceMeters] is the backend `distance_meters` target (a strong
  /// distance signal). [repsSpec] is the raw unit-bearing target string.
  static TrackingMetric resolve({
    required String name,
    String? equipment,
    bool isTimed = false,
    int? holdSeconds,
    int? durationSeconds,
    String? trackingTypeHint,
    num? distanceMeters,
    String? repsSpec,
  }) {
    // 1. Trust the backend hint when it's a known value.
    final hint = _fromHint(trackingTypeHint);
    if (hint != null) return hint;

    final n = name.toLowerCase();
    final eq = (equipment ?? '').toLowerCase();
    final specMetric = repsSpec != null ? parseTarget(repsSpec).metric : null;

    final hasDistanceSignal = (distanceMeters != null && distanceMeters > 0) ||
        specMetric == TrackingMetric.distance;

    // 2. Distance: an explicit distance value/unit, or a distance machine /
    //    sled / carry name. A run/jog/walk counts only with a distance signal
    //    (otherwise it's a timed cardio block).
    if (hasDistanceSignal ||
        _matchesAny(n, _distanceFragments) ||
        (_matchesAny(n, _runFragments) && hasDistanceSignal)) {
      return TrackingMetric.distance;
    }

    // 3. Time: holds / durations / a time-unit target.
    if (isTimed ||
        (holdSeconds != null && holdSeconds > 0) ||
        (durationSeconds != null && durationSeconds > 0) ||
        specMetric == TrackingMetric.time ||
        _matchesAny(n, _timeFragments)) {
      return TrackingMetric.time;
    }

    // 4. Bodyweight reps.
    if (_bodyweightEquipment.contains(eq) ||
        _matchesAny(n, _bodyweightFragments)) {
      return TrackingMetric.bodyweight;
    }

    // 5. Default: loaded weight × reps.
    return TrackingMetric.weight;
  }

  static TrackingMetric? _fromHint(String? hint) {
    switch (hint?.toLowerCase().trim()) {
      case 'weight':
      case 'weighted':
        return TrackingMetric.weight;
      case 'bodyweight':
      case 'bw':
      case 'reps':
        return TrackingMetric.bodyweight;
      case 'time':
      case 'timed':
      case 'duration':
      case 'hold':
        return TrackingMetric.time;
      case 'distance':
      case 'cardio':
        return TrackingMetric.distance;
      default:
        return null;
    }
  }

  static bool _matchesAny(String haystack, List<String> fragments) {
    for (final f in fragments) {
      if (haystack.contains(f)) return true;
    }
    return false;
  }

  /// Parse a free-text target into a metric + magnitude.
  /// Examples: "1000 m"→(distance,1000) · "1 km"→(distance,1000) ·
  /// "8 minutes"→(time,480) · "45s hold"→(time,45) · "100 reps"→(bodyweight,100).
  /// Falls back to (weight, parsedInt?) for a plain number.
  static TargetSpec parseTarget(String raw) {
    final s = raw.toLowerCase().trim();
    // Pull the leading number (handles "1000 m", "1.5 km", "30s").
    final numMatch = RegExp(r'(\d+(?:\.\d+)?)').firstMatch(s);
    final value = numMatch != null ? num.tryParse(numMatch.group(1)!) : null;

    // Distance.
    if (s.contains('km') || s.contains('kilome')) {
      return TargetSpec(TrackingMetric.distance,
          value != null ? (value * 1000).round() : null);
    }
    if (RegExp(r'\bm\b').hasMatch(s) ||
        s.contains('meter') ||
        s.contains('metre') ||
        s.endsWith('m')) {
      // Guard against "min"/"minute" being caught by endsWith('m').
      if (!s.contains('min')) {
        return TargetSpec(TrackingMetric.distance, value?.round());
      }
    }

    // Time.
    if (s.contains('hour') || s.contains('hr')) {
      return TargetSpec(
          TrackingMetric.time, value != null ? (value * 3600).round() : null);
    }
    if (s.contains('min')) {
      return TargetSpec(
          TrackingMetric.time, value != null ? (value * 60).round() : null);
    }
    if (s.contains('sec') ||
        RegExp(r'\d+\s*s\b').hasMatch(s) ||
        s.contains('hold')) {
      return TargetSpec(TrackingMetric.time, value?.round());
    }

    // Reps (bodyweight intent).
    if (s.contains('rep')) {
      return TargetSpec(TrackingMetric.bodyweight, value?.round());
    }

    // Plain number → assume reps for a weighted/strength prescription.
    return TargetSpec(TrackingMetric.weight, value?.round());
  }

  // ── Loaded carries (weight + distance/time) ─────────────────────────────
  /// Loaded sleds / carries whose PRIMARY metric is distance/time but which
  /// ALSO take a load — when true, a distance/time exercise additionally shows
  /// an editable weight column. Mirrors backend `_LOADED_*` in
  /// services/exercise_tracking_metric.py.
  static const List<String> _loadedCarryFragments = [
    'sled', 'prowler', 'yoke', 'farmer', 'suitcase carry', 'loaded carry',
    'overhead carry', 'waiter walk', 'sandbag carry', 'sandbag lunge',
  ];
  static const Set<String> _loadedEquipment = {
    'sled', 'sandbag', 'dumbbell', 'dumbbells', 'kettlebell', 'kettlebells',
    'barbell', 'weight plate', 'trap bar', 'yoke',
  };

  /// True when a distance/time move is loaded (sled, prowler, yoke, carry,
  /// sandbag, or a loaded-equipment token).
  static bool isLoadedCarry(String name, String? equipment) {
    final n = name.toLowerCase();
    for (final f in _loadedCarryFragments) {
      if (n.contains(f)) return true;
    }
    final eq = (equipment ?? '').toLowerCase();
    return _loadedEquipment.contains(eq);
  }

  /// Resolve the full capability profile (ordered metric columns + primary).
  /// Prefers an explicit backend `metric_keys` list; otherwise derives from the
  /// single-metric [resolve] + the loaded-carry rule. Distance/time stay the
  /// primary metric; `weight` is added for loaded carries.
  static TrackingProfile resolveProfile({
    required String name,
    String? equipment,
    bool isTimed = false,
    int? holdSeconds,
    int? durationSeconds,
    String? trackingTypeHint,
    num? distanceMeters,
    String? repsSpec,
    List<String>? explicitKeys,
  }) {
    if (explicitKeys != null && explicitKeys.isNotEmpty) {
      return TrackingProfile(
        List<String>.from(explicitKeys),
        _primaryFor(explicitKeys),
      );
    }
    final metric = resolve(
      name: name,
      equipment: equipment,
      isTimed: isTimed,
      holdSeconds: holdSeconds,
      durationSeconds: durationSeconds,
      trackingTypeHint: trackingTypeHint,
      distanceMeters: distanceMeters,
      repsSpec: repsSpec,
    );
    final keys = <String>[];
    switch (metric) {
      case TrackingMetric.weight:
        keys.addAll(['weight', 'reps']);
        break;
      case TrackingMetric.bodyweight:
        keys.add('reps');
        break;
      case TrackingMetric.time:
        keys.add('time');
        break;
      case TrackingMetric.distance:
        keys.add('distance');
        break;
    }
    // Loaded carry: a distance/time station that also takes a load.
    if ((metric == TrackingMetric.distance || metric == TrackingMetric.time) &&
        isLoadedCarry(name, equipment)) {
      keys.insert(0, 'weight');
    }
    final primary = switch (metric) {
      TrackingMetric.distance => 'distance',
      TrackingMetric.time => 'time',
      _ => 'reps',
    };
    return TrackingProfile(keys, primary);
  }

  /// Headline metric from an explicit key list: first non-weight key, else first.
  static String _primaryFor(List<String> keys) {
    for (final k in keys) {
      if (k != 'weight') return k;
    }
    return keys.isNotEmpty ? keys.first : 'reps';
  }
}

/// The resolved capability profile of an exercise: the ordered metric columns
/// it logs, plus the headline (`primary`) metric. Built by
/// [ExerciseTrackingMetric.resolveProfile].
class TrackingProfile {
  final List<String> metricKeys;
  final String primary;
  const TrackingProfile(this.metricKeys, this.primary);

  bool tracks(String key) => metricKeys.contains(key);
  bool get tracksWeight => metricKeys.contains('weight');
  bool get tracksReps => metricKeys.contains('reps');
  bool get tracksDistance => metricKeys.contains('distance');
  bool get tracksTime => metricKeys.contains('time');
}

/// One metric definition in the shared registry. Mirrors backend
/// services/metric_registry.py — keep the `bagKey` strings identical.
class MetricDef {
  final String key;
  final String bagKey; // JSON key in performance_logs.metrics
  final String label;
  final String shortLabel; // set-table column header
  final String canonicalUnit;
  final String input; // 'number' | 'integer' | 'duration'
  final bool firstClass; // mirrors to a typed performance_logs column
  const MetricDef(this.key, this.bagKey, this.label, this.shortLabel,
      this.canonicalUnit, this.input, this.firstClass);
}

/// Built-in metric catalog (extensible; user-custom metrics merge on top).
const Map<String, MetricDef> kMetricCatalog = {
  'weight': MetricDef('weight', 'weight_kg', 'Weight', 'KG', 'kg', 'number', true),
  'reps': MetricDef('reps', 'reps', 'Reps', 'REPS', 'count', 'integer', true),
  'distance': MetricDef('distance', 'distance_m', 'Distance', 'DIST', 'm', 'number', true),
  'time': MetricDef('time', 'time_s', 'Time', 'TIME', 's', 'duration', true),
  'box_height': MetricDef('box_height', 'box_height_cm', 'Box Height', 'HT', 'cm', 'number', false),
  'calories': MetricDef('calories', 'calories', 'Calories', 'CAL', 'kcal', 'number', false),
  'incline': MetricDef('incline', 'incline_pct', 'Incline', 'INC', 'pct', 'number', false),
  'speed': MetricDef('speed', 'speed_kmh', 'Speed', 'SPD', 'kmh', 'number', false),
  'rpm': MetricDef('rpm', 'rpm', 'RPM', 'RPM', 'rpm', 'integer', false),
  'height': MetricDef('height', 'height_cm', 'Height', 'HT', 'cm', 'number', false),
};
